p2p/src/behaviour/peer_manager.rs

```diff
@@ -170,3 +203,143 @@ impl NetworkBehaviour for PeerManager {
         Ok(dummy::ConnectionHandler)
     }
 }
+
+pub struct BootstrapPeerManager {
+    bootstrap_peers: AHashMap<PeerId, BootstrapPeer>,
+    metrics: Metrics,
+}
+
+impl Debug for BootstrapPeerManager {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
+        f.debug_struct("BootstrapPeerManager")
+            .field("bootstrap_peers", &self.bootstrap_peers)
+            .finish()
+    }
+}
+
+pub struct BootstrapPeer {
+    multiaddr: Multiaddr,
+    dial_backoff: ExponentialBackoff,
+    dial_future: Option<BoxFuture<'static, ()>>,
+}
+
+impl Debug for BootstrapPeer {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
+        f.debug_struct("BootstrapPeer")
+            .field("multiaddr", &self.multiaddr)
+            .field("dial_backoff", &self.dial_backoff)
+            .finish()
+    }
+}
+
+impl BootstrapPeerManager {
+    fn new(bootstrap_peers: &[Multiaddr], metrics: Metrics) -> Result<Self> {
+        let bootstrap_peers: Result<AHashMap<PeerId, BootstrapPeer>, anyhow::Error> =
+            bootstrap_peers
+                .iter()
+                .map(|multiaddr| {
+                    let mut addr = multiaddr.to_owned();
+                    if let Some(Protocol::P2p(peer_id)) = addr.pop() {
+                        Ok((peer_id, BootstrapPeer::new(multiaddr.to_owned())))
+                    } else {
+                        Err(anyhow!("Could not parse bootstrap addr {}", multiaddr))
+                    }
+                })
+                .collect();
```

Comment:
nit, it is common to do `.collect::< Result<AHashMap<PeerId, BootstrapPeer>, anyhow::Error>>()?;` This way the `?` operator is as close to the Result as possible.